pkg/registry/installer.go

```diff
@@ -0,0 +1,252 @@
+// Copyright © 2022 Alibaba Group Holding Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package registry
+
+import (
+	"fmt"
+	"net"
+	"path/filepath"
+	"strconv"
+
+	"github.com/sealerio/sealer/common"
+	"github.com/sealerio/sealer/pkg/clustercert/cert"
+	"github.com/sealerio/sealer/pkg/imagedistributor"
+	"github.com/sealerio/sealer/pkg/infradriver"
+	v2 "github.com/sealerio/sealer/types/api/v2"
+	netutils "github.com/sealerio/sealer/utils/net"
+	osutils "github.com/sealerio/sealer/utils/os"
+	"github.com/sealerio/sealer/utils/strings"
+	"github.com/sirupsen/logrus"
+)
+
+// Installer provide registry lifecycle management.
+type Installer interface {
+	// Reconcile registry deploy hosts thought comparing current deploy host and desiredHosts and return the final registry deploy hosts.
+	// if current deploy host is less than desiredHosts , means scale-up registry node
+	// if current deploy host is bigger than desiredHosts , means scale-down registry node
+	// if current deploy host is equal targetHosts , do nothing
+	// launch registry node
+	// scale-up registry node
+	// scale down registry node
+	Reconcile(desiredHosts []net.IP) ([]net.IP, error)
+
+	// Clean all registry deploy hosts
+	Clean() error
+}
+
+func NewInstaller(currentDeployHost []net.IP,
+	regConfig *v2.LocalRegistry,
+	infraDriver infradriver.InfraDriver,
+	distributor imagedistributor.Distributor) Installer {
+	return &localInstaller{
+		currentDeployHosts: currentDeployHost,
+		infraDriver:        infraDriver,
+		LocalRegistry:      regConfig,
+		distributor:        distributor,
+	}
+}
+
+type localInstaller struct {
```

Comment:
localInstaller and local registry configurator, difference? why need them both?

Reply:
Installer provide registry lifecycle management local registry will implement it. Configurator provide registry configuration management, local registry and external registry will implement it.